kernel/objects/rapl-driver-intel/objects/RaplDriverIntel.cc

```diff
@@ -0,0 +1,254 @@
+        case CPU_KNIGHTS_MILL:
+          dram_avail=true;
+          different_units=true;
+          break;
+
+        case CPU_SANDYBRIDGE:
+        case CPU_IVYBRIDGE:
+          pp0_avail=true;
+          pp1_avail=true;
+          break;
+
+        case CPU_HASWELL:
+        case CPU_HASWELL_ULT:
+        case CPU_HASWELL_GT3E:
+        case CPU_BROADWELL:
+        case CPU_BROADWELL_GT3E:
+        case CPU_ATOM_GOLDMONT:
+        case CPU_ATOM_GEMINI_LAKE:
+        case CPU_ATOM_DENVERTON:
+          pp0_avail=true;
+          pp1_avail=true;
+          dram_avail=true;
+          break;
+
+        case CPU_SKYLAKE:
+        case CPU_SKYLAKE_HS:
+        case CPU_KABYLAKE:
+        case CPU_KABYLAKE_MOBILE:
+        //case CPU_COFFEELAKE:
+        //case CPU_COFFEELAKE_U:
+          pp0_avail=true;
+          pp1_avail=true;
+          dram_avail=true;
+          psys_avail=true;
+          break;
+        default:
+          MLOG_ERROR(mlog::boot, "Unknown CPU model :(");
+      }
+    
+    }else if(cpu_fam == 11){
+      switch(cpu_model){
+
+        case CPU_KNIGHTS_CORNER:
+          MLOG_ERROR(mlog::boot, "Please find out which RAPL sensors are supported for KNC...");
+          break;
+        default:
+          MLOG_ERROR(mlog::boot, "Unknown CPU model :(");
+      }
+    }else{
+      MLOG_ERROR(mlog::boot, "Unsupported processor family");
+      return;
+    }
+
+    //determine power unit
+    //power_units = std::pow(0.5, bits(x86::getMSR(MSR_RAPL_POWER_UNIT),3,0));
+    power_units = bits(x86::getMSR(MSR_RAPL_POWER_UNIT),3,0);
+    MLOG_INFO(mlog::boot, "msr_rapl_power_units", power_units);
+
+    //determine time unit
+    //time_units = std::pow(0.5, bits(x86::getMSR(MSR_RAPL_POWER_UNIT),19,16));
+    time_units = bits(x86::getMSR(MSR_RAPL_POWER_UNIT),19,16);
+    MLOG_INFO(mlog::boot, "time_units", time_units);
+
+    //determine cpu energy unit
+    //cpu_energy_units = std::pow(0.5, bits(x86::getMSR(MSR_RAPL_POWER_UNIT),12,8));
+    cpu_energy_units = bits(x86::getMSR(MSR_RAPL_POWER_UNIT),12,8);
+    MLOG_INFO(mlog::boot, "cpu_energy_units", cpu_energy_units);
+
+    if(different_units){
+      //dram_energy_units = std::pow(0.5, 16.0);
+      dram_energy_units = 16;
+    }else{
+      dram_energy_units = cpu_energy_units; 
+    }
+    MLOG_INFO(mlog::boot, "dram_energy_units", dram_energy_units);
+    
+    //printEnergy();
+  }
+
+  void RaplDriverIntel::invoke(Tasklet* t, Cap self, IInvocation* msg)
+  {
+    monitor.request(t, [=](Tasklet* t){
+        Error err = Error::NOT_IMPLEMENTED;
+        switch (msg->getProtocol()) {
+        case protocol::RaplDriverIntel::proto:
+          err = protocol::RaplDriverIntel::dispatchRequest(this, msg->getMethod(), t, self, msg);
+          break;
+        }
+        if (err != Error::INHIBIT) {
+          msg->replyResponse(err);
+          monitor.requestDone();
+        }
+      } );
+  }
+
+  void RaplDriverIntel::printEnergy(){
+    if(!isIntel) return;
```

Comment:
not needed because then pp0_avail etc are all false

Reply:
done